src/libmeasurement_kit/ooni/meek_fronting.cpp

```diff
@@ -0,0 +1,67 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include "../common/utils.hpp"
+#include <measurement_kit/ooni.hpp>
+
+namespace mk {
+namespace ooni {
+
+using namespace mk::report;
+
+void meek_fronting(std::string input, Settings options,
+                   Callback<Var<report::Entry>> callback,
+                   Var<Reactor> reactor, Var<Logger> logger) {
+    Var<Entry> entry(new Entry);
+
+    std::list<std::string> outer_inner = split(input, ":");
+    if (outer_inner.size() != 2) {
+        return; /* XXX use more specific error */
+    }
+    // XXX: We should make sure that we remove leading and trailing whitespaces
+    // url parsing methods require a schema
+    std::string outer_host = "https://" + outer_inner.front();
+    std::string inner_host = "https://" + outer_inner.back();
+
+    ErrorOr<http::Url> outer_url = mk::http::parse_url_noexcept(outer_host);
+    ErrorOr<http::Url> inner_url = mk::http::parse_url_noexcept(inner_host);
+
+    if (!outer_url || !inner_url) {
+        logger->debug("Invalid url: '%s' or '%s'", outer_host.c_str(),
+                      inner_host.c_str());
+        callback(entry);
```

Comment:
Here I think it would be nicer to do:

```C++
reactor->call_soon([=]() { callback(entry); });
```

This will *schedule* calling `callback(entry)` during the next I/O loop.

I find it nicer because the caller has the guarantee that the callback will always execute after the function to which he/she passed the callback to has returned.

I also think that probably in this case you would like to set the `failure` of the entry.